### PR TITLE
Workload finalizer

### LINKED ISSUE #518: Record injected node affinity in batch Job
**What would you like to be added**:

When we want to suspend Job, we'd like to restore the original nodeAffinity, but some times we can't find the derived workload, see https://github.com/kubernetes-sigs/kueue/blob/045697c46f307b37a03f57f3b42d11a74fe01927/pkg/controller/workload/job/job_controller.go#L397-L402

I'd like to add the nodeAffinity to Job annotations to make this an accurate one. It would like:
```
annotations: kueue.sig.kubernetes.io/injected-node-affinity: '{"key1": "value1", "key2": "value2"}'
```

**Why is this needed**:

Always make sure that when suspending a Job, we'll restore the original nodeAffinity.

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Add and manage resourceInUse finalizer for workloads, to add the posibility to restore the podSets Info upon workload deletion.

#### Which issue(s) this PR fixes:

Fixes #518

#### Special notes for your reviewer:

This PR is split of #834 ,  the code was reviewed there.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by trasc** [comment]
/hold 
for #834

**Comment by alculquicondor** [comment]
Let's park this for the next release

/hold

**Comment by alculquicondor** [comment]
/assign @mimowo
for a review pass

File: pkg/controller/jobframework/reconciler.go [hunk 1]
```diff
@@ -103,14 +104,31 @@ func NewReconciler(
...
 	object := job.Object()
-	if err := r.client.Get(ctx, req.NamespacedName, object); err != nil {
-		// we'll ignore not-found errors, since there is nothing to do.
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+	log := ctrl.LoggerFrom(ctx).WithValues("job", req.String())
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	err := r.client.Get(ctx, req.NamespacedName, object)
+
+	if apierrors.IsNotFound(err) || !object.GetDeletionTimestamp().IsZero() {
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:112]: Could you provide maybe first some context for what is the scenario we are fixing here? I was thinking the scenario is that a user deletes a workload, thus losing the node selectors. However, this if appears to be handling deletion of a Job, so I'm not sure how this is connected.

In particular, this code has user-facing impact. Please explain in the release note the impact, the fixed scenario.

Initially, I was thinking there would be a finalizer added to workloads, at workloads creation. Then, there would be a check in the workload_controller to see if the OwnerReference (the Job) exists. If the Job exists, then we can patch the workload to remove the finalizer.

File: pkg/controller/jobframework/reconciler.go [hunk 2]
```diff
@@ -123,7 +141,7 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
 				"queueName", QueueName(job), "parentWorkload", ParentWorkloadName(job))
 			return ctrl.Result{}, nil
 		}
-		isParentJobManaged, err := r.IsParentJobManaged(ctx, job.Object(), namespacedName.Namespace)
+		isParentJobManaged, err := r.IsParentJobManaged(ctx, job.Object(), req.Namespace)
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:144]: why this change?

File: pkg/controller/jobframework/reconciler.go [hunk 3]
```diff
@@ -164,6 +182,32 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
...
+		log.V(2).Info("The workload is marked for deletion")
+		if !job.IsSuspended() {
+			err := r.stopJob(ctx, job, object, wl, "Workload is deleted")
+			if err != nil {
+				log.Error(err, "Suspending job with deleted workload")
+			}
+			return ctrl.Result{}, err
+		} else {
+			// make sure the finalizer is removed
+			err := r.setFinalizer(ctx, wl, false)
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:197]: I would like to have a dedicated function, such as  `removeFinalizer`. Also, would we need to remove the finalizer here, if we added the finalizer removal to the workload_controller, that deletes the finalizer, if the job is suspended?

File: pkg/controller/jobframework/reconciler.go [hunk 4]
```diff
@@ -164,6 +182,32 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, err
 	}
 
+	// 1.1 If the workload is pending deletion, suspend the job if needed
+	// and drop the finalizer.
+	if wl != nil && !wl.DeletionTimestamp.IsZero() {
+		log.V(2).Info("The workload is marked for deletion")
+		if !job.IsSuspended() {
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:189]: I'm not sure we want this check, as it was a source of issue in a recently fixed bug: https://github.com/kubernetes-sigs/kueue/pull/1005

File: pkg/controller/jobframework/reconciler.go [hunk 5]
```diff
@@ -164,6 +182,32 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
...
+			return ctrl.Result{}, err
+		} else {
+			// make sure the finalizer is removed
+			err := r.setFinalizer(ctx, wl, false)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// 1.2 if a job stop failure is suspected
+	if suspectJobStopFailure(job, wl) {
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:203]: This looks rather suspicious. First, why do we need this , and do we need this given https://github.com/kubernetes-sigs/kueue/pull/1005. Second, if really needed add a comment to explain a scenario in which this is needed.

File: pkg/controller/jobframework/reconciler.go [hunk 6]
```diff
@@ -328,6 +372,22 @@ func (r *JobReconciler) ensureOneWorkload(ctx context.Context, job GenericJob, o
 		}
 	}
 
+	// the job and workload might not be equivalent if the previous
+	// reconcile failed mid stopJob, in which case the job is IsSuspended
+	// the workload is not Admitted but it has the finalizer set
+	// search again
+	if match == nil && job.IsSuspended() {
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:379]: Similarly here, is this needed after https://github.com/kubernetes-sigs/kueue/pull/1005?

File: pkg/controller/jobframework/reconciler.go [hunk 7]
```diff
@@ -409,6 +474,11 @@ func (r *JobReconciler) equivalentToWorkload(job GenericJob, object client.Objec
 
 // startJob will unsuspend the job, and also inject the node affinity.
 func (r *JobReconciler) startJob(ctx context.Context, job GenericJob, object client.Object, wl *kueue.Workload) error {
+	err := r.setFinalizer(ctx, wl, true)
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:477]: This will require an extra API request. Why not add the finalizer during workload creation in ConstructWorkload?

File: pkg/controller/jobframework/reconciler.go [hunk 8]
```diff
@@ -435,23 +505,47 @@ func (r *JobReconciler) stopJob(ctx context.Context, job GenericJob, object clie
 		if stoppedNow {
 			r.record.Eventf(object, corev1.EventTypeNormal, "Stopped", eventMsg)
 		}
-		return err
+		if err != nil {
+			return err
+		}
+	} else {
+		if job.IsSuspended() {
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:512]: This `if` is fairly surprising. Again, I suspect it might not be needed after https://github.com/kubernetes-sigs/kueue/pull/1005. Please explain if it is really needed.

File: pkg/controller/jobframework/reconciler.go [hunk 9]
```diff
@@ -435,23 +505,47 @@ func (r *JobReconciler) stopJob(ctx context.Context, job GenericJob, object clie
...
+}
+
+func suspectJobStopFailure(job GenericJob, wl *kueue.Workload) bool {
+	return wl != nil && job.IsSuspended() && !workload.IsAdmitted(wl) && controllerutil.ContainsFinalizer(wl, kueue.ResourceInUseFinalizerName)
+}
 
-	job.Suspend()
-	if info != nil {
-		job.RestorePodSetsInfo(info)
+func (r *JobReconciler) setFinalizer(ctx context.Context, wl *kueue.Workload, on bool) error {
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:536]: I would actually prefer to have 2 functions: addFinalizer and removeFinalizer.

File: test/integration/controller/job/job_controller_test.go [hunk 10]
```diff
@@ -845,7 +846,7 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", ginkgo.Orde
 		util.ExpectAdmittedActiveWorkloadsMetric(prodClusterQ, 1)
 	})
 
-	ginkgo.It("Should unsuspend job iff localQueue is in the same namespace", func() {
+	ginkgo.It("Should unsuspend job if localQueue is in the same namespace", func() {
```
Comment (mimowo) [inline test/integration/controller/job/job_controller_test.go:849]: Is it a drive by fix, or a legitimate change in semantics. If the former, please do a small separate PR.

**Review by mimowo — COMMENTED** [review]
Mostly concerns about complications with handling mid stopped jobs. I hope this can be now simplified after https://github.com/kubernetes-sigs/kueue/pull/1005. Also, wondering if the code can be simpler, if it was workload_controller that removes the workload. Finally, there is an extra API call now to add the finalizer, I would like to add it when creating the workload.

**Comment by trasc** [comment]
was unintentionally closed

File: pkg/controller/jobframework/reconciler.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/jobframework/reconciler.go:112]: I see. Still, we could check if the owner Job exists or is being deleted (as here) in the workload_controller. Needing to deal with only a single owner job seems simpler.

To the second part, keeping the finalizer for the duration of the job existence seems simpler conceptually, and fewer API requests on suspend / unsuspend, so I'm trying to understand the motivation to delete the workload on suspend. Also, I'm not sure about keeping the workload until the  nodeSelectors are restored. How do we know if this happened? I guess we would need to have some condition to indicate that the nodeSelectors are restored. Otherwise relying just on suspend=true may give wrong results. WDYT?

**Comment by mimowo** [comment]
I'm wondering if we can reopen this, so there is not need to transfer the discussion. We probably hit this https://github.com/isaacs/github/issues/361 based on the message `Failed to re-open PR: state cannot be changed. The workload-finalizer branch was force-pushed or recreated.`. Could you try if this works: https://github.com/isaacs/github/issues/361#issuecomment-219088644. If it doesn't it is fine for me, but I've seen this also somewhere else, so it would be nice to have a working solution.

File: pkg/controller/jobframework/reconciler.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/controller/jobframework/reconciler.go:112]: > I guess we would need to have some condition to indicate that the nodeSelectors are restored. Otherwise relying just on suspend=true may give wrong results. 

Given how Job requires 3 different API calls to suspend, it's indeed not enough to rely on the suspend field.
It's cleaner to just keep the finalizer until the Job is deleted (has a deletionTimestamp) or the Workload has a Finished condition.

Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:126]: Initially , the presence of the finalizer for in the workload of a suspended job was an indicator that `stopJob` failed somewhere mid execution. However #1005 has another approach.

Comment (mimowo) [reply pkg/controller/jobframework/reconciler.go:126]: What is the user-facing difference between:
1. workload lifetime is just for the running job
2. workload lifetime tied to the Job lifetime, for as long as they are equivalent

In (1.) we need two additional calls to add and remove the finalizer, so the performance is degraded (it can be argued how much). Are there drawbacks of (2.)?

Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:126]: with 2, deleting a workload while the job is still present requires the job controller to remove the finalizer (one additional reconcile cycle, if triggered from outside the job controller)

Comment (alculquicondor) [reply pkg/controller/jobframework/reconciler.go:126]: We could remove the finalizer from Workload in 2 scenarios:
- The Workload is already marked as finished.
- The Job is deleted.

Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:126]: But still need to be removed if not finished and the job is suspended if the workload gets deleted.

Still the current implementation is the best fit, in my mind.

Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:126]: > In particular, in the scenario of manual workload deletion (after suspending) it requires 2 API calls (one to add and one to delete the finalizer).

I don't follow.

Comment (mimowo) [reply pkg/controller/jobframework/reconciler.go:126]: I had in mind that you require 2 API calls [here](https://github.com/kubernetes-sigs/kueue/blob/46980b39077eb773a08f48c4712c2fac622b0c41/pkg/controller/jobframework/reconciler.go#L474) and [here](https://github.com/kubernetes-sigs/kueue/blob/46980b39077eb773a08f48c4712c2fac622b0c41/pkg/controller/jobframework/reconciler.go#L522).
So, when you mentioned workload deletion I was thinking about the scenario:
1. Job created [here is the first call to add the finalizer]
2. job started
3. job suspended, workload unadmitted [+ API  call to remove the finalizer]
4. workload manually deleted by user, and replacement workload created [+ API call to add the finalizer, and +API call to eventually delete it]

Comment (mimowo) [reply pkg/controller/jobframework/reconciler.go:126]: I guess, you had a different scenario, that the Job is never started, and the workload is manually deleted, right? I see, in this case the method you propose would not require any extra calls vs. one in the alternative. Still, the scenario seems much less important. Anyway, I will defer to @alculquicondor for the decision.

Comment (alculquicondor) [reply pkg/controller/jobframework/reconciler.go:126]: To me, performance (in this case, less API calls in the happy path) and simplicity takes higher priority than covering an edge use case.

Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:126]: The main purpose of this change is to make sure that when a job is unsuspended and its workload gets removed we still get the chance to restore it's node selectors. That is why I choose to "lock" the deletion for the time the job is unsuspended only. (In the previous iterations, before #1005, this approach was also useful in detecting mid stop errors ).

In an ideal life-cycle of a job/workload  (the workload is created once, is admitted and the job finishes) this will have one extra API call then the case when we set the finalizer at the time of creating the workload.  The number of API calls will change in favor of this approach if the workload gets recreated and against is if the workload gets preempted.

However changed so the finalizer is set from the workloads creation until it gets to `Finished`.

File: pkg/controller/jobs/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_controller_test.go:1]: Add new test cases to cover the reconciler changes

File: pkg/controller/jobframework/reconciler.go (hunk 1 above)
Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:126]: > Not really. The finalizer can be included in the initial spec sent to create the Workload. This is not only one less API call, but overall less prone to race conditions.

That is the one I'm talking about, the removal needs to be done in a different call. (if it gets to finish, or if it gets deleted before that).

Comment (alculquicondor) [reply pkg/controller/jobframework/reconciler.go:126]: Right. The point is that in this approach we no longer need an extra API call when admitting a workload.

Comment (trasc) [reply pkg/controller/jobframework/reconciler.go:126]: But need one if/when the workload gets deleted before being admitted (if the job spec changes). anyway....

File: pkg/controller/jobs/job/job_controller_test.go
Comment (trasc) [reply pkg/controller/jobs/job/job_controller_test.go:1]: done, for removal on finished, the removal on workload and job deletion is done in integration.

File: test/integration/controller/job/job_controller_test.go [hunk 11]
```diff
@@ -875,15 +876,30 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", ginkgo.Orde
...
 
-		ginkgo.By("job should be unsuspended")
+		ginkgo.By("job should be unsuspended and NodeSelector properly set")
 		gomega.Eventually(func() *bool {
 			gomega.Expect(k8sClient.Get(ctx, lookupKey, createdProdJob)).Should(gomega.Succeed())
 			return createdProdJob.Spec.Suspend
 		}, util.Timeout, util.Interval).Should(gomega.Equal(ptr.To(false)))
+
+		runningSelector := maps.Clone(createdProdJob.Spec.Template.Spec.NodeSelector)
+		gomega.Expect(runningSelector).NotTo(gomega.Equal(originalSelector))
```
Comment (alculquicondor) [inline test/integration/controller/job/job_controller_test.go:892]: I would just remove this check, as it's redundant given the one below.

File: test/integration/controller/job/job_controller_test.go [hunk 12]
```diff
@@ -875,15 +876,30 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", ginkgo.Orde
...
+
+		expectedSelector := maps.Clone(originalSelector)
+		if expectedSelector == nil {
+			expectedSelector = onDemandFlavor.Spec.NodeLabels
+		} else {
+			for k, v := range onDemandFlavor.Spec.NodeLabels {
+				expectedSelector[k] = v
+			}
+		}
```
Comment (alculquicondor) [inline test/integration/controller/job/job_controller_test.go:901]: avoid generating expected values in tests. hard-coding it is more readable https://enterprisecraftsmanship.com/posts/dry-damp-unit-tests/

File: test/integration/controller/job/job_controller_test.go (hunk 11 above)
Comment (trasc) [reply test/integration/controller/job/job_controller_test.go]: done

File: test/integration/controller/job/job_controller_test.go (hunk 12 above)
Comment (trasc) [reply test/integration/controller/job/job_controller_test.go]: done

File: test/integration/controller/job/job_controller_test.go [hunk 13]
```diff
@@ -945,6 +961,123 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", ginkgo.Orde
...
+					gomega.Expect(k8sClient.Get(ctx, wlKey, wl)).Should(gomega.Succeed())
+					return wl.Finalizers
+				}, util.Timeout, util.Interval).Should(gomega.ContainElement(kueue.ResourceInUseFinalizerName))
+			})
+
+			ginkgo.By("deleting the job", func() {
+				gomega.Expect(k8sClient.Delete(ctx, job)).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("checking the workload is removed or the finalizer is no longer set", func() {
```
[the remark this answers is not in the document]
Comment (trasc) [reply test/integration/controller/job/job_controller_test.go]: In fact I think we should focus on the finalizer being removed, since the deletion of the workload may or may not be trigger based on the `DeletePropagationPolicy`.

**Comment by alculquicondor** [comment]
Let's hope the new default merge method works
